Publish conda build artifacts on Azure as pipeline artifacts.

ryanvolz opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Added a ``news`` entry

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

This adds steps to the Azure pipeline templates that save the conda build output directory as an Azure pipeline artifact. This is useful for inspecting the build result and testing built packages before they are published. I think I've set it up so that the artifact publishing won't run if the directory doesn't exist, and also that it will still run even if the build fails (making the work directory available, for example). I tested this on one recipe, https://github.com/conda-forge/soapysdr-module-rtlsdr-feedstock/pull/1, and it works at least when the builds succeed. Wider testing would probably be wise.

This is an evolution of a similar feature that has already been implemented for conda-forge/staged-recipes (https://github.com/conda-forge/staged-recipes/pull/11827, https://github.com/conda-forge/staged-recipes/pull/11832).

I don't know if there need to be considerations for the size of the published artifacts, or how long they are kept around. As far as I can tell, the artifacts will exist for as long as Microsoft allows the pipeline to keep them around.

beckermr commented:
cc @chrisburr who had some ideas on this that might generalize to other platforms. I am not opposed to this PR, but I would like you all to coordinate efforts

isuruf commented:
I think we should ask azure (or find an article) about best practices if we are going to enable this for all feedstocks.

isuruf commented:
See https://docs.microsoft.com/en-us/azure/devops/artifacts/concepts/best-practices?view=azure-devops
and https://docs.microsoft.com/en-us/azure/devops/artifacts/reference/limits?view=azure-devops

I don't what the package ID mentioned in the last article is

ryanvolz commented:
What I don't really understand, despite having put this together, is that the pipeline artifacts used here are somehow different than the artifacts that most of the Azure docs talk about. This doesn't do package publishing, add there is no feed from which to consume these artifacts, as those docs talk about. Clarification from Azure or someone who knows better would be great.

jaimergp commented:
> I think we should ask azure (or find an article) about best practices if we are going to enable this for all feedstocks.

Maybe opt-in via `conda-forge.yml` first?

beckermr commented:
We can set a short lifetime for these in the devop UI so that we are not using a ton of storage as well.

ryanvolz commented:
> > I think we should ask azure (or find an article) about best practices if we are going to enable this for all feedstocks.
> 
> Maybe opt-in via `conda-forge.yml` first?

I think this is a good idea, and was going to come here and suggest it myself to get this moving again before you beat me to the punch. :-P I can make the publishing conditional in the templates as is done with other parts of the template, but I don't immediately know how the variables make it to the template from `conda-forge.yml` and if I need to make anything aware that there is a new variable. Can anyone provide a little direction before I try to dig through the code?
 
> > ... pipeline artifacts used here are somehow different than the artifacts that most of the Azure docs talk about.
> 
> Afaik, by using `publish:`, we invoke a task named `PublishPipelineArtifact@1`, which uses [Pipeline Artifacts](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/publish-pipeline-artifact?view=azure-devops) -- these are different from the plain "Azure Artifacts" concept the docs Isuru linked mention, I think. More kinds of artifacts are listed [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/artifacts-overview?view=azure-devops#supported-artifact-types).
> 
> There's a subtle message about the storage limits in [this FAQ](https://docs.microsoft.com/en-us/azure/devops/artifacts/start-using-azure-artifacts?toc=%2Fazure%2Fdevops%2Forganizations%2Ftoc.json&bc=%2Fazure%2Fdevops%2Forganizations%2Fbreadcrumb%2Ftoc.json&view=azure-devops#q-which-artifacts-count-toward-my-total-billed-storage), namely:
> 
> > You are not billed by Azure Artifacts for storage of Pipeline Artifacts, Build Artifacts, and Pipeline Caching.

Thanks for looking that up! I thought I had read these things somewhere, but couldn't immediately find it again for reference.

> The [docs here](https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml#limiting-which-files-are-included) also recommend this trick for the artifact name:
> 
> > Deleting and/or overwriting Pipeline Artifacts is not currently supported. The recommended workflow if you want to re-run a failed pipeline job is to include the job ID in the artifact name. `$(system.JobId)` is the appropriate variable for this purpose. See System variables to learn more about predefined variables.

I can easily tack the job ID on to the end of the name and it seems a sensible thing to do. It would now be `conda_artifacts_$(CONFIG)_$(system.JobId)`. Would swapping job ID and config (arch) be better for sorting purposes, if that's ever relevant?

> Regarding [`.artifactignore`](https://docs.microsoft.com/en-us/azure/devops/artifacts/reference/artifactignore?view=azure-devops), it could be used to only publish `tar.bz2` files with this config:
> 
> ```
> **/*
> !*.tar.bz2
> ```

I can do this if desired as well, but I think it was @beckermr who commented on the staged-recipes change that the entirety of the conda `build_artifacts` directory can be useful for debugging purposes. Is there anything from the directory that really is useless and worth excluding? Picking up *.bz2 instead of only *.tar.bz2 looks like it would bring in the various forms of `repodata.json`, and maybe that captures everything useful?

> We can set a short lifetime for these in the devop UI so that we are not using a ton of storage as well.

Definitely needed, I think, and something that we can test live in small quantities by making the artifact publishing opt-in through `conda-forge.yml`.

beckermr commented:
The only thing you may not care about is some html stuff in that dir. I'd keep the whole thing since it is mostly packages anyways.

jaimergp commented:
In case somebody wants to see how the artifact ID `conda_artifacts_$(CONFIG)_$(system.JobId)` would look like:

https://dev.azure.com/openmm/feedstock-builds/_build/results?buildId=391&view=artifacts&type=publishedArtifacts

beckermr commented:
Can we include the lifetime in the config, or do we have to set that in the ui?

ryanvolz commented:
> Can we include the lifetime in the config, or do we have to set that in the ui?

If you're asking me, I haven't found any documentation on how to set the lifetime of the pipeline artifacts, so I don't know how it is possible. If you've found an option in the UI, then that's more than I've got.

Other than that, the only remaining question on this is whether it is desirable to make this opt-in through `conda-forge.yml`. I have no problem with doing that, but also don't know *how* to do that.

scopatz commented:
@ryanvolz - I think you need to change `configure_feedstock.py` to add new keys

ryanvolz commented:
I added a `conda-forge.yml` key called `store_build_artifacts` under the `azure` flag to toggle this, defaulting to False. I did a quick test locally and re-rendered a recipe without modifying `conda-forge.yml` and the pipeline artifacts were skipped, then I added `azure: store_build_artifacts: true` and re-rendered and had the pipeline artifacts included. So this looks good to go from my perspective.

beckermr commented:
We need to make sure to set a short time limit in the azure UI. @conda-forge/core what about 1 week for now?

scopatz commented:
1 week is probably enough

beckermr commented:
OK. I went ahead and set this to 14 days for now.

beckermr commented:
@conda-forge/core ready to merge?

scopatz commented:
Please put in a PR to the docs for this new parameter: https://conda-forge.org/docs/maintainer/conda_forge_yml.html

ryanvolz commented:
> Please put in a PR to the docs for this new parameter: https://conda-forge.org/docs/maintainer/conda_forge_yml.html

I just put in the docs PR.
